vsphere/internal/helper/virtualmachine/virtual_machine_helper.go

```diff
@@ -75,6 +89,60 @@ func FromUUID(client *govmomi.Client, uuid string) (*object.VirtualMachine, erro
 	return vm.(*object.VirtualMachine), nil
 }
 
+// virtualMachineFromSearchIndex gets the virtual machine reference via the
+// SearchIndex MO and is the method used to fetch UUIDs on newer versions of
+// vSphere.
+func virtualMachineFromSearchIndex(ctx context.Context, client *govmomi.Client, uuid string) (object.Reference, error) {
+	log.Printf("[DEBUG] Using SearchIndex to look up UUID %q", uuid)
+	search := object.NewSearchIndex(client.Client)
+	result, err := search.FindByUuid(ctx, nil, uuid, true, structure.BoolPtr(false))
+	if err != nil {
+		return nil, err
+	}
+
+	if result == nil {
+		return nil, newUUIDNotFoundError(fmt.Sprintf("virtual machine with UUID %q not found", uuid))
+	}
+
+	return result, nil
+}
+
+// virtualMachineFromContainerView is a compatability method that is
+// used when the version of vSphere is too old to support using SearchIndex's
+// FindByUuid method correctly. This is mainly to facilitate the ability to use
+// FromUUID to find both templates in addition to virtual machines, which
+// historically was not supported by FindByUuid.
+func virtualMachineFromContainerView(ctx context.Context, client *govmomi.Client, uuid string) (object.Reference, error) {
+	log.Printf("[DEBUG] Using ContainerView to look up UUID %q", uuid)
+	m := view.NewManager(client.Client)
+
+	v, err := m.CreateContainerView(ctx, client.ServiceContent.RootFolder, []string{"VirtualMachine"}, true)
+	if err != nil {
+		return nil, err
+	}
+
+	defer func() {
+		if err = v.Destroy(ctx); err != nil {
+			panic(err)
```

Comment:
if you would rather not panic here, you could just wrap the body in an inner func, and do this afterwards, but seems fine to me as is, as its probably an edge.